### ISSUE TITLE
Can't restore containers

### ISSUE THREAD

**Issue by boucher** [moby/moby#35691]
Description

Restoring a container no longer works in the latest version of moby/docker.

Steps to reproduce the issue:

*** Before this works, you may need to upgrade CRIU to a newer version than the one in the standard shell Dockerfile ***

```
export CRIU_VERSION=3.6; apt-get install libnet-dev -y && \
	mkdir -p /usr/src/criu \
	&& curl -sSL https://github.com/xemul/criu/archive/v${CRIU_VERSION}.tar.gz | tar -v -C /usr/src/criu/ -xz --strip-components=1 \
	&& cd /usr/src/criu \
	&& make \
	&& make install-criu
```

Create a container:
`docker run --name crtest -d --security-opt seccomp:unconfined busybox           /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'`

Create a checkpoint of that container:
`docker checkpoint create crtest checkpoint1`

Stop the container (see #35690):
`docker stop crtest`

Attempt to restore from the checkpoint:
`docker start --checkpoint checkpoint1 crtest`

Additional information you deem important (e.g. issue happens only occasionally):

The error I see output in the logs is this:

```
ERRO[2017-12-04T18:00:14.593051694Z] 6f75781052e4b95e9f829345bff7bb42365a33f2d06578b84be288a89efa316c cleanup: failed to delete container from containerd: no such container 
ERRO[2017-12-04T18:00:14.593360378Z] Handler for POST /v1.30/containers/6f/start returned error: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/6f75781052e4b95e9f829345bff7bb42365a33f2d06578b84be288a89efa316c/init.pid: no such file or directory: unknown 
Error response from daemon: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/6f75781052e4b95e9f829345bff7bb42365a33f2d06578b84be288a89efa316c/init.pid: no such file or directory: unknown
```

Output of docker version:
```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:15:15 2017
 OS/Arch:      linux/amd64

Server:
 Version:      dev
 API version:  1.35 (minimum version 1.12)
 Go version:   go1.9.2
 Git commit:   7ef9266
 Built:        Mon Dec  4 17:07:37 2017
 OS/Arch:      linux/amd64
 Experimental: true
Output of docker info:

Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: dev
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6bff39c643886dfa3d546e83a90a527b64ddeacf
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.49-moby
Operating System: Debian GNU/Linux 9 (stretch) (containerized)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.855GiB
Name: ddaf0d20627d
ID: NRZK:4RHA:M4ZF:TAVO:ROAJ:AZQZ:YDMD:ZW4J:TEQD:RTBN:XK6F:RC65
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 34
 System Time: 2017-12-04T17:32:49.110767545Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Additional environment details (AWS, VirtualBox, physical, etc.):
```
Docker for Mac
Version 17.09.0-ce-mac35 (19611)
Channel: stable
a98b7c1b7c
```

**Comment by thaJeztah** [comment]
ping @kolyshkin @mlaventure

**Comment by danedb** [comment]
I have the same problem.
Ubuntu 17.10 Linux version 4.13.0-31-generic

All was well and working in docker ce 17.03 then I upgraded to 17.12 and it broke.

**Comment by kolyshkin** [comment]
I'll work on that

**Comment by pgayvallet** [comment]
While trying to add the feature in docker for desktop, I reproduced the issue with the scenario given by the author.

```
$ docker run --name crtest -d --security-opt seccomp:unconfined nginx
2c24c3ee2a1e5c6067be4ae25d6aa67868109af0587eade62a2757d9a9ea665d

$ docker checkpoint create crtest checkpoint1
checkpoint1

$ docker stop crtest
crtest

$ docker start --checkpoint checkpoint1 crtest
Error response from daemon: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/2c24c3ee2a1e5c6067be4ae25d6aa67868109af0587eade62a2757d9a9ea665d/init.pid: no such file or directory: unknown
```

Also, after having created a checkpoint, listing them also fails because of another missing file : 

```sh
docker checkpoint ls crtest
Error response from daemon: open /var/lib/docker/containers/2c24c3ee2a1e5c6067be4ae25d6aa67868109af0587eade62a2757d9a9ea665d/checkpoints/checkpoint1/config.json: no such file or directory
```

docker version (18.03.0-ce-rc2-mac56)
```
docker version
Client:
 Version:	18.03.0-ce-rc3
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	e730959
 Built:	Thu Mar  8 20:12:09 2018
 OS/Arch:	darwin/amd64
 Experimental:	true
 Orchestrator:	kubernetes

Server:
 Engine:
  Version:	18.03.0-ce-rc3
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	e730959
  Built:	Thu Mar  8 20:20:22 2018
  OS/Arch:	linux/amd64
  Experimental:	true
```

Content of the checkpoint folder : 
```sh
/var/lib/docker/containers/2c24c3ee2a1e5c6067be4ae25d6aa67868109af0587eade62a2757d9a9ea665d/checkpoints/checkpoint1# ls -l
total 200
-rw-r--r-- 1 root root  1788 Mar 13 15:05 core-1.img
-rw-r--r-- 1 root root  1743 Mar 13 15:05 core-28.img
-rw-r-xr-x 1 root root    46 Mar 13 15:05 descriptors.json
-rw-r--r-- 1 root root    44 Mar 13 15:05 fdinfo-2.img
-rw-r--r-- 1 root root    44 Mar 13 15:05 fdinfo-3.img
-rw-r--r-- 1 root root   187 Mar 13 15:05 files.img
-rw-r--r-- 1 root root    18 Mar 13 15:05 fs-1.img
-rw-r--r-- 1 root root    18 Mar 13 15:05 fs-28.img
-rw-r--r-- 1 root root    34 Mar 13 15:05 ids-1.img
-rw-r--r-- 1 root root    34 Mar 13 15:05 ids-28.img
-rw-r--r-- 1 root root    82 Mar 13 15:05 ipcns-var-10.img
-rw-r--r-- 1 root root   439 Mar 13 15:05 mm-1.img
-rw-r--r-- 1 root root   439 Mar 13 15:05 mm-28.img
-rw-r--r-- 1 root root  4168 Mar 13 15:05 mountpoints-12.img
-rw-r--r-- 1 root root   490 Mar 13 15:05 netdev-9.img
-rw-r--r-- 1 root root   898 Mar 13 15:05 netns-9.img
-rw-r--r-- 1 root root    98 Mar 13 15:05 pagemap-1.img
-rw-r--r-- 1 root root    98 Mar 13 15:05 pagemap-28.img
-rw-r--r-- 1 root root 57344 Mar 13 15:05 pages-1.img
-rw-r--r-- 1 root root 45056 Mar 13 15:05 pages-2.img
-rw-r--r-- 1 root root    22 Mar 13 15:05 pipes-data.img
-rw-r--r-- 1 root root    36 Mar 13 15:05 pstree.img
-rw-r--r-- 1 root root   384 Mar 13 15:05 tmpfs-dev-102.tar.gz.img
-rw-r--r-- 1 root root   297 Mar 13 15:05 tmpfs-dev-105.tar.gz.img
-rw-r--r-- 1 root root    98 Mar 13 15:05 tmpfs-dev-106.tar.gz.img
-rw-r--r-- 1 root root    34 Mar 13 15:05 utsns-11.img
```

**Comment by iprimo** [comment]
Hi, 
I am experiencing similar problem creating a Clone container.
error "...failed to retrieve OCI runtime container pid..."

Details in the following:


_+_+_ Goal

```
docker run -d --name looper2 --security-opt seccomp:unconfined busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'

docker checkpoint create looper2 checkpoint2

docker create --name looper-clone --security-opt seccomp:unconfined busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'

copy -r /source-container/checkpoint /clone-container/

docker start --checkpoint=checkpoint2 looper-clone
```

_+_+_ Result

```
24504c853ff4 -> source container
34ad1fdbeced -> cloned container

root@abc:/home/usr1/folder# ls /var/lib/docker/containers/24504...../
24504c853ff4c138e46c27f4193fbaf7dc1563c6117c899c6d1eaf59f86ed24a-json.log  checkpoints  config.v2.json  hostconfig.json  hostname  hosts  resolv.conf  resolv.conf.hash  shm

root@abc:/home/usr1/folder# ls /var/lib/docker/containers/34ad1fdbecedf...../
34ad1fdbecedf6159a6712bcd29e360ffb223e3b6dcc3b4e758239d7807ca5c0-json.log  checkpoints  config.v2.json  hostconfig.json  hostname  hosts  resolv.conf  resolv.conf.hash  shm

root@abc:/home/usr1/folder# docker start --checkpoint=checkpoint2 looper-clone

Error response from daemon: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/34ad1fdbecedf6159a6712bcd29e360ffb223e3b6dcc3b4e758239d7807ca5c0/init.pid: no such file or directory: unknown

root@abc:/home/usr1/folder# docker ps -a | grep looper
34ad1fdbeced        busybox             "/bin/sh -c 'i=0; wh…"   7 minutes ago       Created                                          looper-clone
24504c853ff4        busybox             "/bin/sh -c 'i=0; wh…"   7 minutes ago       Up 7 minutes                                     looper2
root@abc:/home/usr1/folder#
```

_+_+_ Versioning 

```
root@abc:/home/usr1/folder# docker version
Client:
 Version: 17.12.1-ce
 API version:  1.35
 Go version:   go1.9.4
 Git commit:   7390fc6
 Built:   Tue Feb 27 22:17:53 2018
 OS/Arch: linux/amd64

Server:
 Engine:
  Version:     17.12.1-ce
  API version: 1.35 (minimum version 1.12)
  Go version:  go1.9.4
  Git commit:  7390fc6
  Built:  Tue Feb 27 22:16:25 2018
  OS/Arch:     linux/amd64
  Experimental:     true


root@abc:/home/usr1/folder# criu --version
Version: 3.4

root@abc:/home/usr1/folder# criu check
Looks good.

root@abc:/home/usr1/folder# lsb_release -a
No LSB modules are available.
Distributor ID:     Ubuntu
Description:   Ubuntu 17.10
Release:  17.10
Codename: artful
root@abc:/home/usr1/folder# uname -r
4.13.0-36-generic

root@abc:/home/usr1/folder#runc 
.....
VERSION:
   1.0.0-rc5+dev
```

**Comment by saeidscorp** [comment]
Hi,
Same here.
Any progress on knowing the main cause of the issue?

**Comment by pascalwhoop** [comment]
Has anyone upgraded to 18.03 and seen any changes? I can't fetch the newer version from their repo atm, I guess I'm on stable or something.

**Comment by cpuguy83** [comment]
There haven't been any changes to checkpoint/restore yet.

**Comment by pascalwhoop** [comment]
Running this with 17.06.0-ce works fine when restoring the same way as described in [this post](https://technology.amis.nl/2018/04/08/first-steps-with-docker-checkpoint-to-create-and-restore-snapshots-of-running-containers/) which is a simple nodejs app.

**Comment by avagin** [comment]
I'm working on this issue: ￼
https://github.com/containerd/containerd/pull/2425

**Comment by avagin** [comment]
This issue was fixed in runc:
https://github.com/opencontainers/runc/commit/8187fb740c202f5d29f1717bb933143c10dae8a1

**Comment by tswift242** [comment]
@avagin has that fix gotten into moby yet?

**Comment by avagin** [comment]
@tswift242 As far as I know, it isn't there yet. I asked @kolyshkin to help with updating runc in moby.

**Comment by xuchenCN** [comment]
LGTM + 1

**Comment by kolyshkin** [comment]
So, it's complicated... here's a dependency chain:

I have opened https://github.com/moby/moby/pull/37568 but closed it in favor of https://github.com/moby/moby/pull/37710 which also updates runc.

Unfortunately that one can't be merged until https://github.com/opencontainers/runc/issues/1837 is fixed (it is being fixed in https://github.com/opencontainers/runc/pull/1862 right now).

**Comment by kolyshkin** [comment]
OK to fix this we need:
* moby commit db621eb7ee6a837f68 (PR #37360) :heavy_check_mark: merged
* runc commit 8187fb740c202f5d2 (PR https://github.com/opencontainers/runc/pull/1840) :heavy_check_mark: merged
* update runc in moby (PR #37710) :x: TODO

Same changes need to be backported to docker-ce/ee releases if we want to fix those.

PS there's also https://github.com/containerd/containerd/pull/2425 but looks like it's not currently required.

**Comment by crashtua** [comment]
@kolyshkin is it intended to work with established TCP connections?
Trying to run command "docker checkpoint create node.0.cl1 fresh" but in logs I see
```
(00.366961) 27833 fdinfo 7: pos:                0 flags:                2/0
(00.366964) 	Searching for socket 37fbc8a (family 2.6)
(00.366967) Error (criu/sk-inet.c:181): Connected TCP socket, consider using --tcp-established option.
(00.366973) ----------------------------------------
(00.366985) Error (criu/cr-dump.c:1352): Dump files (pid: 27833) failed with -1
```

btw, it works for simple bash print loop :)

**Comment by crashtua** [comment]
@kolyshkin any comments for my message above?

**Comment by zhendong2050** [comment]
It seems that this issues is not solved in Docer 18.03-ce. I got the similar error when I restored a checkpoint:

 docker start --checkpoint checkpoint0 cr
Error response from daemon: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/9703a036e4fac25544ec08bd65fb87a6aaa6731d46ce9de656f0366abf3c4746/init.pid: no such file or directory: unknown

Versions:
Ubuntu 18.04
Docker version 18.06.1-ce, build e68fc7a
Criu Version: 3.10

**Comment by kolyshkin** [comment]
> is it intended to work with established TCP connections?

@crashtua CRIU supports that, I'm not sure why `--tcp-establised` flag is not set. It is supported by both runc and containerd (`AllowOpenTCP`) but there's no way to set it from docker (cli/daemon).

**Comment by tswift242** [comment]
@kolyshkin https://github.com/moby/moby/pull/37710 was merged so this should be fixed in master?

**Comment by kolyshkin** [comment]
> so this should be fixed in master?

Indeed it is @tswift242! I have just checked it (with runc binary from git master) and it works!

**Comment by kolyshkin** [comment]
@thaJeztah I guess this should be closed )

**Comment by thaJeztah** [comment]
:tada:

**Comment by crashtua** [comment]
@kolyshkin looks like that options created in https://github.com/moby/moby/blob/master/libcontainerd/client_daemon.go#L562 and only Exit set to true here. I will try to enable everything here and will see if container state completely preserved. Still, not sure if this will work, since container after that will be recreated, since all virtual network adapters will be new devices, so I suspect that CRIU will fail to restore processes states...

**Comment by xuchenCN** [comment]
Client:
 Version:           18.06.3-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        d7080c1
 Built:             Wed Feb 20 02:26:51 2019
 OS/Arch:           linux/amd64
 Experimental:      true

Server:
 Engine:
  Version:          18.06.3-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       d7080c1
  Built:            Wed Feb 20 02:28:17 2019
  OS/Arch:          linux/amd64
  Experimental:     true

uanme -a
Linux 3.10.0-693.el7.x86_64 #1 SMP Tue Aug 22 21:09:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

criu --version
Version: 3.11
GitID: v3.11-80-gf3f4ed5

Error response from daemon: failed to retrieve OCI runtime container pid: open /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/fa8159cf9a7a2064a3d4ff4fc37a5828625a533ff5299e965911c26150ab03a3/init.pid: no such file or directory: unknown

This issue already fixed now ?

**Comment by avagin** [comment]
https://github.com/moby/moby/pull/37710 was merged  on Oct 5, 2018, so the docker version where this issue is fixed should be higher that 18.10.

**Comment by buck2202** [comment]
Just FYI...this recently started happening to me on 18.09.08 on ubuntu server 16.04 LTS installations, possibly after a criu update, but I'm not totally sure what the trigger was. It wasn't a uniform issue with 16.04 LTS, though, because I had identical docker+criu versions running on 16.04 google cloud instances without issue.

The issue resolved when I moved from the ga kernel (4.4.x) to the hwe kernel (4.15.x); the google cloud image for server-16.04 was apparently on hwe by default.
`sudo apt install --install-recommends linux-generic-hwe-16.04`
See https://wiki.ubuntu.com/Kernel/LTSEnablementStack


Just adding a note here for anyone else googling after error messages

**Comment by avagin** [comment]
> this recently started happening to me on 18.09.8 on ubuntu server 16.04 LTS installations,

Could you describe what is "this"? What version of CRIU do you use? Could you attache criu logs?

**Comment by buck2202** [comment]
Hi, sorry, I was referring to the error message reported in the initial post and from @xuchenCN immediately above. Containers would fail to start from checkpoints with messages like
```Error response from daemon: failed to retrieve OCI runtime container pid: open <long path>/init.pid: no such file or directory: unknown```

My CRIU version is 3.13 from the PPA here (https://launchpad.net/~criu/+archive/ubuntu/ppa). I've migrated all of my affected machines/VMs off of the 4.4 kernel and am not seeing the issue now, but if you're still interested in tracking it down I can spin one back up.

**Comment by cnnrznn** [comment]
> > this recently started happening to me on 18.09.8 on ubuntu server 16.04 LTS installations,
> 
> Could you describe what is "this"? What version of CRIU do you use? Could you attache criu logs?

I'm having an issue doing c/r with some docker containers. I want to work on custom checkpoint dirs but I can't find where Docker/Moby/runc/containerd/God stores the CRIU logs. Any pointers?

Thanks :)

**Comment by buck2202** [comment]
> > > this recently started happening to me on 18.09.8 on ubuntu server 16.04 LTS installations,
> > 
> > 
> > Could you describe what is "this"? What version of CRIU do you use? Could you attache criu logs?
> 
> I'm having an issue doing c/r with some docker containers. I want to work on custom checkpoint dirs but I can't find where Docker/Moby/runc/containerd/God stores the CRIU logs. Any pointers?
> 
> Thanks :)

Probably off-topic here, but...to my knowledge, creating checkpoints to custom dirs is fine, but restoring from them is broken (see moby/moby#37344). I workaround it by just wrapping restores in a bash script that preps a custom-dir restore by copying the relevant checkpoint to the default location (in my case, `/var/lib/docker/containers/ID/checkpoints/`) and then deleting it after the container has restored. see https://github.com/moby/moby/issues/37344#issuecomment-450782189

If you mean you want to work on the issue, maybe the other bug report would give some guidance.

**Comment by cnnrznn** [comment]
I saw that issue and tried following the comment. I'll leave a question on that thread with versions/problem. 

Thanks!

**Comment by cryokev** [comment]
> @kolyshkin any comments for my message above?

if any one reading this is experiencing this, i found a helpful article that fixes the issue.

>I started looking for a way to enable tcp-establish. The CRIU configuration guide suggested echo 'tcp-established' > /etc/criu/runc.conf for containerized deployments. However, doing it had no effect. That’s when I found that the support for configs only arrived in runc 1.0-rc7 and CRIU 3.11 - while Ubuntu packages came with older runc 1.0-rc4 and CRIU 3.6.

source: https://kirshatrov.com/2019/07/06/experimenting-with-criu-and-slow-boot

TLDR: build latest criu from source.
